units/provisioning_create_host.go

```diff
@@ -73,66 +90,97 @@ func (j *createHostJob) Run(ctx context.Context) {
+			MaxTime: j.host.SpawnOptions.TimeoutSetup.Sub(j.start),
+		})
 	}
 
-	settings := j.env.Settings()
-
-	err = j.createHost(ctx, j.host, settings)
-	j.AddError(err)
+	j.AddError(j.createHost(ctx))
 }
 
-func (j *createHostJob) createHost(ctx context.Context, h *host.Host, settings *evergreen.Settings) error {
-	hostStartTime := time.Now()
+func (j *createHostJob) createHost(ctx context.Context) error {
+	if err := ctx.Err(); err != nil {
+		return errors.Wrap(err, "canceling create host because context is canceled")
+	}
+	hostStartTime := j.start
 	grip.Info(message.Fields{
-		"message": "attempting to start host",
-		"hostid":  h.Id,
-		"job":     j.ID(),
+		"message":      "attempting to start host",
+		"hostid":       j.host.Id,
+		"job":          j.ID(),
+		"attempt":      j.CurrentAttempt,
+		"max_attempts": j.MaxAttempts,
 	})
 
-	cloudManager, err := cloud.GetManager(ctx, h.Provider, settings)
+	cloudManager, err := cloud.GetManager(ctx, j.host.Provider, j.env.Settings())
 	if err != nil {
 		grip.Warning(message.WrapError(err, message.Fields{
 			"message": "problem getting cloud provider for host",
-			"host":    h.Id,
+			"host":    j.host.Id,
 			"job":     j.ID(),
 		}))
-		return errors.Wrapf(errIgnorableCreateHost, "problem getting cloud provider for host '%s' [%s]", h.Id, err.Error())
+		return errors.Wrapf(errIgnorableCreateHost, "problem getting cloud provider for host '%s' [%s]", j.host.Id, err.Error())
 	}
 
-	if err = h.Remove(); err != nil {
-		grip.Notice(message.WrapError(err, message.Fields{
-			"message": "problem removing intent host",
-			"job":     j.ID(),
-			"host":    h.Id,
-		}))
-		return errors.Wrapf(errIgnorableCreateHost, "problem getting cloud provider for host '%s' [%s]", h.Id, err.Error())
+	// On the first attempt, remove the intent document so no other create host job tries to create this intent.
+	if j.CurrentAttempt == 1 {
+		if err := j.host.Remove(); err != nil {
+			grip.Notice(message.WrapError(err, message.Fields{
+				"message": "problem removing intent host",
+				"job":     j.ID(),
+				"host":    j.host.Id,
+			}))
+			return errors.Wrapf(errIgnorableCreateHost, "problem removing intent host '%s' [%s]", j.host.Id, err.Error())
+		}
 	}
 
-	_, err = cloudManager.SpawnHost(ctx, h)
-	if err != nil {
-		return errors.Wrapf(err, "error spawning host %s", h.Id)
+	defer j.tryRequeue(ctx)
+	if _, err = cloudManager.SpawnHost(ctx, j.host); err != nil {
+		return errors.Wrapf(err, "error spawning host %s", j.host.Id)
 	}
 
-	h.Status = evergreen.HostStarting
+	j.host.Status = evergreen.HostStarting
 
-	// Provisionally set h.StartTime to now. Cloud providers may override
+	// Provisionally set j.host.StartTime to now. Cloud providers may override
 	// this value with the time the host was created.
-	h.StartTime = time.Now()
+	j.host.StartTime = j.start
 
-	_, err = h.Upsert()
-	if err != nil {
-		return errors.Wrapf(err, "error updating host %v", h.Id)
+	if err := j.host.Insert(); err != nil {
```

Comment:
Won't this hit duplicate key errors?

Reply:
No, if spawning a host fails, this function returns on 139. It only inserts the host once, i.e., if it succeeds.

Reply:
there's a db.IsDuplicateKey(err error) function that you can use to return nil in this case, which is probably what you want (for belt and suspenders, we should also change that function to use errors.Cause()